TombEngine/Objects/TR3/Entity/tr3_civvy.cpp

```diff
@@ -79,341 +119,484 @@ namespace TEN::Entities::Creatures::TR3
-					extraTorsoRot.y = AI.angle;
-				}
+				case CIVVY_STATE_AIM_FAR_PUNCH:
 
-				if (AI.bite && AI.distance < CIVVY_ATTACK2_RANGE)
-					item->Animation.TargetState = CIVVY_PUNCH2;
-				else
-					item->Animation.TargetState = CIVVY_STATE_WALK_FORWARD;
+					creature.MaxTurn = CIVVY_AIM_TURN_RATE_MAX;
+					creature.Flags = 0;
 
-				break;
+					if (AI.ahead)
+					{
+						jointTorsoRot.x = AI.xAngle;
+						jointTorsoRot.y = AI.angle;
+					}
 
-			case CIVVY_PUNCH0:
-				creature->MaxTurn = CIVVY_WALK_TURN_RATE_MAX;
+					if ((creature.Enemy->IsLara() && AI.ahead && AI.distance < CIVVY_ATTACK_FAR_PUNCH_RANGE) ||
+						(creature.Enemy->IsCreature() && creature.Enemy->HitPoints > 0 && AI.ahead && AI.distance < SQUARE(BLOCK(0.5f))) )
+						item.Animation.TargetState = CIVVY_STATE_ATTACK_FAR_PUNCH;
+					else
+						item.Animation.TargetState = CIVVY_STATE_IDLE;
 
-				if (AI.ahead)
-				{
-					extraTorsoRot.x = AI.xAngle;
-					extraTorsoRot.y = AI.angle;
-				}
+					break;
 
-				if (!creature->Flags && item->TouchBits.Test(CivvyAttackJoints))
-				{
-					CreatureEffect(item, CivvyBite, DoBloodSplat);
-					DoDamage(creature->Enemy, CIVVY_ATTACK_DAMAGE);
-					SoundEffect(SFX_TR4_LARA_THUD, &item->Pose);
-					creature->Flags = 1;
-				}
+				case CIVVY_STATE_AIM_WALKING_PUNCH:
 
-				break;
+					creature.MaxTurn = CIVVY_AIM_TURN_RATE_MAX;
+					creature.Flags = 0;
 
-			case CIVVY_PUNCH1:
-				creature->MaxTurn = CIVVY_WALK_TURN_RATE_MAX;
+					if (AI.ahead)
+					{
+						jointTorsoRot.x = AI.xAngle;
+						jointTorsoRot.y = AI.angle;
+					}
 
-				if (AI.ahead)
-				{
-					extraTorsoRot.x = AI.xAngle;
-					extraTorsoRot.y = AI.angle;
-				}
+					
+					if ((creature.Enemy->IsLara() && AI.bite && AI.distance < CIVVY_ATTACK_WALKING_PUNCH_RANGE) ||
+						(creature.Enemy->IsCreature() && creature.Enemy->HitPoints > 0 && AI.ahead && AI.distance < SQUARE(BLOCK(0.5f))) )
+						item.Animation.TargetState = CIVVY_STATE_ATTACK_WALKING_PUNCH;
+					else
+						item.Animation.TargetState = CIVVY_STATE_WALK;
 
-				if (!creature->Flags && item->TouchBits.Test(CivvyAttackJoints))
-				{
-					CreatureEffect(item, CivvyBite, DoBloodSplat);
-					DoDamage(creature->Enemy, CIVVY_ATTACK_DAMAGE);
-					SoundEffect(SFX_TR4_LARA_THUD, &item->Pose);
-					creature->Flags = 1;
-				}
+					break;
 
-				if (AI.ahead && AI.distance > CIVVY_ATTACK1_RANGE && AI.distance < CIVVY_ATTACK2_RANGE)
-					item->Animation.TargetState = CIVVY_PUNCH2;
+				case CIVVY_STATE_ATTACK_CLOSE_PUNCH:
 
-				break;
+					creature.MaxTurn = CIVVY_WALK_TURN_RATE_MAX;
 
-			case CIVVY_PUNCH2:
-				creature->MaxTurn = CIVVY_WALK_TURN_RATE_MAX;
+					if (AI.ahead)
+					{
+						jointTorsoRot.x = AI.xAngle;
+						jointTorsoRot.y = AI.angle;
+					}
 
-				if (AI.ahead)
-				{
-					extraTorsoRot.x = AI.xAngle;
-					extraTorsoRot.y = AI.angle;
-				}
+					if (creature.Flags == 0 &&
+						item.Animation.AnimNumber == GetAnimNumber(item, CIVVY_ANIM_CLOSE_PUNCH_ATTACK))
```

Comment:
No need for animation check here, official source don't have it either.

Reply:
Done